src/NUnitEngine/nunit.engine/Services/AgentStore.cs

```diff
@@ -10,45 +10,57 @@ namespace NUnit.Engine.Services
     /// <summary>
     /// Defines the agent tracking operations that must be done atomically.
     /// </summary>
-    internal sealed partial class AgentStore
+    internal sealed class AgentStore
     {
-        private readonly Dictionary<Guid, AgentRecord> _agentsById = new Dictionary<Guid, AgentRecord>();
+        private static readonly Logger log = InternalTrace.GetLogger(typeof(AgentStore));
+
+        private readonly object LOCK = new object();
+
+        private readonly Dictionary<Guid, AgentRecord> _agentIndex = new Dictionary<Guid, AgentRecord>();
+        private readonly Dictionary<Process, AgentRecord> _processIndex = new Dictionary<Process, AgentRecord>();
 
         public void AddAgent(Guid agentId, Process process)
         {
-            lock (_agentsById)
+            Guard.ArgumentNotNull(process, nameof(process));
+
+            lock (LOCK)
             {
-                if (_agentsById.ContainsKey(agentId))
+                if (_agentIndex.ContainsKey(agentId))
                 {
                     throw new ArgumentException($"An agent has already been started with the ID '{agentId}'.", nameof(agentId));
                 }
 
-                _agentsById.Add(agentId, AgentRecord.Starting(process));
+                _agentIndex[agentId] = _processIndex[process] = new AgentRecord(agentId, process);
             }
         }
 
         public void Register(ITestAgent agent)
         {
-            lock (_agentsById)
+            lock (LOCK)
             {
-                if (!_agentsById.TryGetValue(agent.Id, out var record)
+                log.Debug($"Registering agent {agent.Id:B}");
+
+                if (!_agentIndex.TryGetValue(agent.Id, out var record)
                     || record.Status != AgentStatus.Starting)
                 {
-                    throw new ArgumentException($"Agent {agent.Id} must have a status of {AgentStatus.Starting} in order to register, but the status was {record.Status}.", nameof(agent));
+                    string status = record?.Status.ToString() ?? "unknown";
+                    throw new ArgumentException($"Agent {agent.Id} must have a status of {AgentStatus.Starting} in order to register, but the status was {status}.", nameof(agent));
                 }
 
-                _agentsById[agent.Id] = record.Ready(agent);
+                record.Agent = agent;
+                record.Status = AgentStatus.Ready;
             }
         }
 
         public bool IsReady(Guid agentId, [NotNullWhen(true)] out ITestAgent? agent)
         {
-            lock (_agentsById)
+            lock (LOCK)
             {
-                if (_agentsById.TryGetValue(agentId, out var record) && record.IsReady)
+                if (_agentIndex.TryGetValue(agentId, out var record)
+                    && record.Status == AgentStatus.Ready)
                 {
                     agent = record.Agent;
-                    return true;
+                    return agent is object;
```

Comment:
I strongly dislike `is object` to test for `not null`.
```suggestion
                    return agent is not null;
```

Reply:
Done